### PR TITLE
raft: add raft test suite

### PR DESCRIPTION
**Pull request by xiang90**
start a basic raft test suite.

the goal of this pr is to set up the desired network interface and the functionality of node.

a node should be able to restart, pause.
a network should be able to delay, drop message.

### CODE REVIEW

File: raft/rafttest/network.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+type network interface {
+	send(m raftpb.Message)
+	recv() chan raftpb.Message
+	// drop message at given rate (1.0 drops all messages)
+	drop(from, to uint64, rate float64)
+	// delay message for (0, d] randomly at given rate (1.0 delay all messages)
+	// do we need rate here?
+	delay(from, to uint64, d time.Duration, rate float64)
```
Comment (bdarnell) [inline raft/rafttest/network.go]: Maybe a `duplicate(from, to uint64, d time.Duration, rate float64)` option as well? (messages are sent a second time after (0, d] seconds)

Comment (xiang90) [inline raft/rafttest/network.go]: sure.

Comment (bdarnell) [inline raft/rafttest/network.go]: That's true, I guess all the duplicate-message scenarios I can come up with are actually the application retrying.

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**